### ISSUE TITLE
Use with @babel/register and ts-node/register in non-webpack environment

### ISSUE THREAD

**Issue by bensaufley** [zeit/styled-jsx#434]
#### Do you want to request a *feature* or report a *bug*?

🐞Bug

#### What is the current behavior?

**tl;dr: I can't get `styled-jsx` to work with Mocha and TypeScript.**

Some of that I've documented over in #90, and I've had a PR (https://github.com/DefinitelyTyped/DefinitelyTyped/pull/24771) merged on DefinitelyTyped.

I've got a private repo for a work project that is currently using `next@5.1.0`, `@zeit/next-typescript`, and `styled-jsx`. It worked in browser, but in Mocha, which was run using `ts-node/register`, it threw for `styled-jsx/css` ("untranspiled"). At this point I spun up a basic example repo you can see [here](https://github.com/bensaufley/styled-jsx-example-issues) to map my debugging in a public format. As it turns out, of course, the "untranspile" error was expected, since babel was not a part of the Mocha workflow. `babel-register` at version 6 [is not built to be a part of a pipeline](https://github.com/TypeStrong/ts-node/issues/51) and wouldn't accept incoming code from `ts-node/register`, so that meant it simply wasn't an option.

Enter `next@6.0.0canary`, which uses Babel 7. `@babel/register@^7` [uses pirates](https://github.com/babel/babel/pull/3670) which allows `ts-node/register` to pass its output to Babel. And yet…

My first problem can be reproduced on [this commit](https://github.com/bensaufley/styled-jsx-example-issues/commit/e275a9c1d673e48d129f1a59a98320432376b901):

```
/path/to/next-styled-jsx/node_modules/@babel/core/node_modules/source-map/lib/util.js:206
  while (aPath.indexOf(aRoot + '/') !== 0) {
               ^
TypeError: Cannot read property 'indexOf' of undefined
    at Object.relative (/path/to/next-styled-jsx/node_modules/@babel/core/node_modules/source-map/lib/util.js:206:16)
    at BasicSourceMapConsumer.SourceMapConsumer_generatedPositionFor [as generatedPositionFor] (/path/to/next-styled-jsx/node_modules/@babel/core/node_modules/source-map/lib/source-map-consumer.js:731:21)
    at /path/to/next-styled-jsx/node_modules/@babel/core/lib/transformation/file/generate.js:108:47
    at Array.forEach (<anonymous>)
    at BasicSourceMapConsumer.SourceMapConsumer_eachMapping [as eachMapping] (/path/to/next-styled-jsx/node_modules/@babel/core/node_modules/source-map/lib/source-map-consumer.js:155:14)
    at mergeSourceMap (/path/to/next-styled-jsx/node_modules/@babel/core/lib/transformation/file/generate.js:107:20)
    at generateCode (/path/to/next-styled-jsx/node_modules/@babel/core/lib/transformation/file/generate.js:82:17)
    at runSync (/path/to/next-styled-jsx/node_modules/@babel/core/lib/transformation/index.js:38:58)
    at transformSync (/path/to/next-styled-jsx/node_modules/@babel/core/lib/transform-sync.js:15:38)
    at Object.transform (/path/to/next-styled-jsx/node_modules/@babel/core/lib/transform.js:20:65)
    at compile (/path/to/next-styled-jsx/node_modules/@babel/register/lib/node.js:75:20)
    at compileHook (/path/to/next-styled-jsx/node_modules/@babel/register/lib/node.js:104:12)
    at Module._compile (/path/to/next-styled-jsx/node_modules/pirates/lib/index.js:86:27)
    at Module.m._compile (/path/to/next-styled-jsx/node_modules/ts-node/src/index.ts:402:23)
    … etc
```

It looked like the fact that `styled-jsx` doesn't produce sourceMaps by default was throwing off Babel at some point. I don't know if that's true, but I realized that styled-jsx has a `sourceMaps` option so I just flipped that switch.

So now I'm at [this commit](https://github.com/bensaufley/styled-jsx-example-issues/commit/f8dd407e3c7b6fb4560f80b4555a85e89c8afdba), which is failing here:

```
TypeError: Cannot read property 'indexOf' of undefined
    at Object.relative (/path/to/next-styled-jsx/node_modules/source-map/lib/util.js:206:16)
    at SourceMapGenerator_setSourceContent [as setSourceContent] (/path/to/next-styled-jsx/node_modules/source-map/lib/source-map-generator.js:144:21)
    at makeSourceMapGenerator (/path/to/next-styled-jsx/node_modules/styled-jsx/dist/_utils.js:383:13)
    at processCss (/path/to/next-styled-jsx/node_modules/styled-jsx/dist/_utils.js:482:21)
    at processTaggedTemplateExpression (/path/to/next-styled-jsx/node_modules/styled-jsx/dist/babel-external.js:64:44)
    at /path/to/next-styled-jsx/node_modules/styled-jsx/dist/babel-external.js:151:7
    at Array.forEach (<anonymous>)
    at PluginPass.ImportDeclaration (/path/to/next-styled-jsx/node_modules/styled-jsx/dist/babel-external.js:150:31)
    at newFn (/path/to/next-styled-jsx/node_modules/@babel/traverse/lib/visitors.js:223:21)
    at NodePath._call (/path/to/next-styled-jsx/node_modules/@babel/traverse/lib/path/context.js:64:19)
    at NodePath.call (/path/to/next-styled-jsx/node_modules/@babel/traverse/lib/path/context.js:38:17)
    at NodePath.visit (/path/to/next-styled-jsx/node_modules/@babel/traverse/lib/path/context.js:99:12)
    at TraversalContext.visitQueue (/path/to/next-styled-jsx/node_modules/@babel/traverse/lib/context.js:135:18)
    at TraversalContext.visitMultiple (/path/to/next-styled-jsx/node_modules/@babel/traverse/lib/context.js:89:17)
    at TraversalContext.visit (/path/to/next-styled-jsx/node_modules/@babel/traverse/lib/context.js:174:19)
    at Function.traverse.node (/path/to/next-styled-jsx/node_modules/@babel/traverse/lib/index.js:76:17)
    at NodePath.visit (/path/to/next-styled-jsx/node_modules/@babel/traverse/lib/path/context.js:106:18)
    at TraversalContext.visitQueue (/path/to/next-styled-jsx/node_modules/@babel/traverse/lib/context.js:135:18)
    at TraversalContext.visitSingle (/path/to/next-styled-jsx/node_modules/@babel/traverse/lib/context.js:94:19)
    at TraversalContext.visit (/path/to/next-styled-jsx/node_modules/@babel/traverse/lib/context.js:176:19)
    at Function.traverse.node (/path/to/next-styled-jsx/node_modules/@babel/traverse/lib/index.js:76:17)
    at traverse (/path/to/next-styled-jsx/node_modules/@babel/traverse/lib/index.js:46:12)
    at transformFile (/path/to/next-styled-jsx/node_modules/@babel/core/lib/transformation/index.js:108:27)
    at runSync (/path/to/next-styled-jsx/node_modules/@babel/core/lib/transformation/index.js:35:3)
    at transformSync (/path/to/next-styled-jsx/node_modules/@babel/core/lib/transform-sync.js:15:38)
    at Object.transform (/path/to/next-styled-jsx/node_modules/@babel/core/lib/transform.js:20:65)
    at compile (/path/to/next-styled-jsx/node_modules/@babel/register/lib/node.js:75:20)
    at compileHook (/path/to/next-styled-jsx/node_modules/@babel/register/lib/node.js:104:12)
    at Module._compile (/path/to/next-styled-jsx/node_modules/pirates/lib/index.js:86:27)
    at Module.m._compile (/path/to/next-styled-jsx/node_modules/ts-node/src/index.ts:402:23)
```

Digging a bit here, it looks like `var filename = file.opts.sourceFileName` is returning `undefined` [here](https://github.com/zeit/styled-jsx/blob/051d0d182bace7b9d8db374ad783c54f97512e3d/src/_utils.js#L437). Sticking a console.log there shows me that it is, indeed, undefined. I do have access to `file.opts.filename` at that point. Swapping out that reference gets me this:

```
Error: Invalid mapping: {"generated":{"line":1,"column":0},"original":{"line":2,"column":19},"name":null}
  at SourceMapGenerator_validateMapping [as _validateMapping] (/path/to/next-styled-jsx/node_modules/styled-jsx/node_modules/source-map/lib/source-map-generator.js:298:13)
    at SourceMapGenerator_addMapping [as addMapping] (/path/to/next-styled-jsx/node_modules/styled-jsx/node_modules/source-map/lib/source-map-generator.js:110:12)
```

That appears to have been passed in from [here](https://github.com/zeit/styled-jsx/blob/051d0d182bace7b9d8db374ad783c54f97512e3d/src/lib/style-transform.js#L30), which again… is not passing `name`.

So I'm beginning to suspect that this package simply isn't Babel7 ready? But I'm not sure. Babel plugins are really outside my comfort zone. Could really use some input here.

Interestingly the app linked above works in a browser. My guess is that Next.js and/or Webpack finesse the problems I'm encountering here one way or another.

#### If the current behavior is a bug, please provide the steps to reproduce and possibly a minimal demo or testcase in the form of a Next.js app, CodeSandbox URL or similar

See above

#### What is the expected behavior?

I am able to run `npm test` with valid `styled-jsx` using `mocha`, `ts-node/register`, and `@babel/register`, and nothing breaks.

#### Environment (include versions)

* OS: MacOS 10.13.4
* Browser: No browser. Node v9.8.0
* styled-jsx (version): 2.2.6

#### Did this work in previous versions?

I haven't found a version where this will work.

**Comment by bensaufley** [comment]
Some more updates here: I've tried some stopgap measures to simply remove the CSS for test environments (on the current Next5 branch):

```ts
const shim = (cb) => {
  if (process.env.NODE_ENV === 'test') return '';
  else return cb();
};

// in another file:
import css from 'styled-jsx/css';

import shim from '@lib/shim';

export default shim(() => (
css`
.my-css {
  stuff: 0;
}
`
));
```

This gives me this error, originating at `node_modules/babel-types/lib/definitions/index.js:109:13`:

```
Module build failed: TypeError: /usr/src/project/components/path/styles.ts: Property body[1] of BlockStatement expected node to be of a type ["Statement"] but instead got "ArrowFunctionExpression"
```
---

I tried as well in my TSX, hoping maybe if the imported file was never referenced it would never call `css` and thus never throw.

```tsx
{process.env.NODE_ENV !== 'test' && <style jsx>{styles}</style>}
```

This produces the following warning:

```
Warning: Received `true` for a non-boolean attribute `jsx`.
If you want to write it to the DOM, pass a string instead: jsx="true" or jsx={value.toString()}.
```

And some server-side rendering that looks like this:

```
i<!-- -->n<!-- -->p<!-- -->u<!-- -->t<!-- -->{<!-- -->b<!-- -->o<!-- -->r<!-- -->d<!-- -->e<!-- -->r<!-- -->:<!-- -->1<!-- -->p<!-- -->x<!-- --> <!-- -->s<!-- -->o<!-- -->l<!-- -->i<!-- -->d<!-- --> <!-- -->#<!-- -->a<!-- -->a<!-- -->a<!-- -->;<!-- -->b<!-- -->o<!-- -->r<!-- -->d<!-- -->e<!-- -->r<!-- -->-<!-- -->r<!-- -->a<!-- -->d<!-- -->i<!-- -->u<!-- -->s<!-- -->:<!-- -->0<!-- -->.<!-- -->6<!-- -->e<!-- -->m<!-- -->;<!-- -->b<!-- -->o<!-- -->x<!-- -->-<!-- -->s<!-- -->h<!-- -->a<!-- -->d<!-- -->o<!-- -->w<!-- -->:<!-- -->#<!-- -->c<!-- -->e<!-- -->c<!-- -->e<!-- -->c<!-- -->e<!-- --> <!-- -->0<!-- --> <!-- -->1<!-- -->p<!-- -->x<!-- --> <!-- -->4<!-- -->p<!-- -->x<!-- --> <!-- -->i<!-- -->n<!-- -->s<!-- -->e<!-- -->t<!-- -->;<!-- -->f<!-- -->o<!-- -->n<!-- -->t<!-- -->-<!-- -->f<!-- -->a<!-- -->m<!-- -->i<!-- -->l<!-- -->y<!-- -->:<!-- -->i<!-- -->n<!-- -->h<!-- -->e<!-- -->r<!-- -->i<!-- -->t<!-- -->;<!-- -->f<!-- -->o<!-- -->n<!-- -->t<!-- -->-<!-- -->s<!-- -->i<!-- -->z<!-- -->e<!-- -->:<!-- -->1<!-- -->e<!-- -->m<!-- -->;<!-- -->h<!-- -->e<!-- -->i<!-- -->g<!-- -->h<!-- -->t<!-- -->:<!-- -->2<!-- -->.<!-- -->3<!-- -->1<!-- -->2<!-- -->5<!-- -->e<!-- -->m<!-- -->;<!-- -->m<!-- -->a<!-- -->r<!-- -->g<!-- -->i<!-- -->n<!-- -->:<!-- -->0<!-- -->;<!-- -->p<!-- -->a<!-- -->d<!-- -->d<!-- -->i<!-- -->n<!-- -->g<!-- -->:<!-- -->0<!-- -->.<!-- -->5<!-- -->e<!-- -->m<!-- -->;<!-- -->w<!-- -->i<!-- -->d<!-- -->t<!-- -->h<!-- -->:<!-- -->1<!-- -->0<!-- -->0<!-- -->%<!-- -->;<!-- -->}<!-- -->
```

All of which I've learned means that the styled-jsx preprocessing isn't happening.

**Comment by giuseppeg** [comment]
@bensaufley thank you for investigating this issue and not giving up! But most importantly I am sorry that you had to do this in the first place. 

I am not familiar with `ts-node/register` at all and probably it would take a while before I find the time to understand how it works, debug and come up with a fix that makes it work with Babel 6. At this point I am not familiar with Babel 7 either yet.

Anyway from your report it seems that, even though sourceMaps are disabled via plugin option we still register the Stylis plugin. I need to double check but that could be a bug and we should rather do something like [this](https://github.com/zeit/styled-jsx/blob/051d0d182bace7b9d8db374ad783c54f97512e3d/src/lib/style-transform.js#L100-L105).

Generally I suggest to remove `styled-jsx/babel` from the tests config. Imo transpiling styled-jsx is useless for test purposes.

I am not super familiar with Mocha but can you mock entire modules? I know you can do that with Jest see https://kevinjalbert.com/jest-snapshots-reducing-styled-jsx-noise/

**Comment by bensaufley** [comment]
@giuseppeg thanks for your response.

For what it's worth, after more thought I kind of doubt Babel 7 is the issue because the local browser environment works fine with Next 6 canary and Babel 7. This should at least indicate that it _can_ work with Babel 7.

> Generally I suggest to remove `styled-jsx/babel` from the tests config. Imo transpiling styled-jsx is useless for test purposes.

In the short term I've been trying to just keep the styled-jsx files out of our tests because they don't work, but in the long term I really think that getting the test code to reflect the live code as much as possible is worthwhile. In the case of something like `styled-jsx`, getting the transpile to be a part of the test flow would mean that if we're passing bad data to `styled-jsx`, or returning bad data from it to our JSX, our tests will blow up.

Basically in the same way that a basic JSX Component test simply ensures that it renders properly and without errors (`expect(mount(<Component />)).to.exist`), a test of our styles code would make sure that importing the file gives us whatever `css` is expected to give.

As for mocks and stubs, we use `Sinon`, but it can't stub a default export. I've spent a short amount of time trying to find to some other sort of register that could implement something like you mentioned Jest supports, but I really feel like that's solving the wrong problem.

FWIW I have found a stopgap solution, which is using the same logic as in my `shim` up above, but using it inline on every styles file rather than as a module:

```ts
export default (() => {
  if (process.env.NODE_ENV === 'test') return '';

  return css`
    .my-css {
      color: blue;
    }
  `;
})();
```

But again, this is a stopgap solution. I feel like there's got to be a way to get `@babel/register` to do what it's expected to do with this plugin.

Thanks again for your help! Please let me know if there's anything I can do to help figure out a solution; this is kind of blocking more work for me at this point.

**Comment by giuseppeg** [comment]
if you can't find a way to mock the file you could use something like that https://www.npmjs.com/package/babel-plugin-module-resolver to alias your mock module to `styled-jsx/css` only in tests (using babel's envs). I will try your testcase (repo) when I have time and see if I can find a solution.

**Comment by bensaufley** [comment]
Follow-up: another issue with stubbing `styled-jsx/css` (using something like [mock-require](https://github.com/boblauer/mock-require)—which I'm now using in my real repo, for the time being) to return a plaintext string is that our tests still output ``Warning: Received `true` for a non-boolean attribute `jsx`.`` when we render a styled component, which may be safely ignorable but generally I'm of the opinion that warnings should be handled rather than ignored whenever possible, and it creates noise that makes it harder to see real issues.

**Comment by athrunsun** [comment]
@bensaufley Hi. Could you provide some instructions on how to add type definitions for `css.resolve` and `css.global` mentioned [here](https://github.com/zeit/styled-jsx#external-styles)?

**Comment by bensaufley** [comment]
@athrunsun ah the new release finally happened! Cool. We've moved to `styled-components` since I last posted here, but I'd imagine what you'd do is something like:

```js
mockRequire('styled-jsx/css', {
  default: () => '',
  global: () => '',
  resolve: () => ({ className: 'resolved-class-name', styles: '' }),
});
```

**Comment by athrunsun** [comment]
@bensaufley Thank you for your instructions! It worked perfectly.

I'll submit a PR to DefinitelyTyped repo.

**Comment by giuseppeg** [comment]
> to return a plaintext string is that our tests still output Warning: Received `true` for a non-boolean attribute `jsx`.

we shipped a solution to this in v3.1.0 https://github.com/zeit/styled-jsx#rendering-in-tests
